proxy/src/main.rs

```diff
@@ -45,52 +45,122 @@ async fn main() -> anyhow::Result<()> {
+                .unwrap()
+                .parse()?;
+            let endpoint = http::Endpoint::new(url, reqwest::Client::new());
+            auth::BackendType::Console(Cow::Owned(endpoint), ())
+        }
+        "postgres" => {
+            let url = arg_matches
+                .get_one::<String>("auth-endpoint")
+                .unwrap()
+                .parse()?;
+            auth::BackendType::Postgres(Cow::Owned(url), ())
+        }
+        "link" => {
+            let url = arg_matches.get_one::<String>("uri").unwrap().parse()?;
+            auth::BackendType::Link(Cow::Owned(url))
+        }
+        other => bail!("unsupported auth backend: {other}"),
+    };
+
+    let config: &ProxyConfig = Box::leak(Box::new(ProxyConfig {
+        tls_config,
+        auth_backend,
+    }));
+
+    info!("Version: {GIT_VERSION}");
+    info!("Authentication backend: {}", config.auth_backend);
+
+    // Check that we can bind to address before further initialization
+    info!("Starting http on {http_address}");
+    let http_listener = TcpListener::bind(http_address).await?.into_std()?;
+
+    info!("Starting mgmt on {mgmt_address}");
+    let mgmt_listener = TcpListener::bind(mgmt_address).await?.into_std()?;
+
+    info!("Starting proxy on {proxy_address}");
+    let proxy_listener = TcpListener::bind(proxy_address).await?;
+
+    let tasks = [
+        tokio::spawn(http::server::task_main(http_listener)),
+        tokio::spawn(proxy::task_main(config, proxy_listener)),
+        tokio::task::spawn_blocking(move || mgmt::thread_main(mgmt_listener)),
+    ]
+    .map(flatten_err);
+
+    set_build_info_metric(GIT_VERSION);
+    // This will block until all tasks have completed.
+    // Furthermore, the first one to fail will cancel the rest.
+    let _: Vec<()> = futures::future::try_join_all(tasks).await?;
+
+    Ok(())
+}
+
+fn cli() -> clap::Command {
+    clap::Command::new("Neon proxy/router")
+        .disable_help_flag(true)
         .version(GIT_VERSION)
         .arg(
             Arg::new("proxy")
                 .short('p')
                 .long("proxy")
-                .takes_value(true)
                 .help("listen for incoming client connections on ip:port")
                 .default_value("127.0.0.1:4432"),
         )
         .arg(
             Arg::new("auth-backend")
                 .long("auth-backend")
-                .takes_value(true)
-                .possible_values(["console", "postgres", "link"])
+                .value_parser(["console", "postgres", "link"])
                 .default_value("link"),
         )
         .arg(
             Arg::new("mgmt")
                 .short('m')
                 .long("mgmt")
-                .takes_value(true)
                 .help("listen for management callback connection on ip:port")
                 .default_value("127.0.0.1:7000"),
         )
         .arg(
             Arg::new("http")
-                .short('h')
```

Comment:
cc @funbringer 

I've removed this parameter due to the new `cli` test started failing below.
It complained, that this `-h` conflicts with the short form of `--help`.

LMK if there's anything else you want to do about this.

Reply:
Sure! To be honest, I'd drop al shorthands.

Reply:
I've reached my breaking change limit recently and prefer to do less of those for a while, so will leave the others as is so far, thank you for confirming 🙂